File: src/inifile.py
import os

class ReadIniFile:
    def __init__(self, config_file: str):
        self._entries = {} # type: dict[str, str]
        
        if not os.path.exists(config_file):
            return

        try:
            with open(config_file, 'r', encoding='utf-8', errors='ignore') as f:
                in_section = ""
                for line in f:
                    # getline in C++ removes the delimiter (newline), so we rstrip it.
                    line = line.rstrip('\n')
                    
                    if not line:
                        continue
                    
                    # Ignore comment lines - strict check at index 0 as in C++
                    if line[0] == '#' or line[0] == ';':
                        continue
                    
                    # A new section
                    if line[0] == '[':
                        end_pos = line.find(']')
                        if end_pos != -1:
                            # C++: substr(1, find(']') - 1);
                            # find returns index, so length is index - 1.
                            # substr(start, length).
                            # Python slice: [start : start + length] -> [1 : 1 + (end_pos - 1)] -> [1 : end_pos]
                            in_section = line[1:end_pos]
                        continue
                    
                    # Get a value
                    pos_equal = line.find('=')
                    if pos_equal != -1:
                        name = line[:pos_equal]
                        value = line[pos_equal+1:]
                        
                        # Strip whitespace from beginning and end
                        name = name.strip(" \t")
                        value = value.strip(" \t")
                        
                        # Insert entry
                        self._entries[f"{in_section}/{name}"] = value
        except IOError:
            # If file cannot be opened, we just return (as C++ does)
            pass

    def get_float(self, section: str, entry: str, default_value: float) -> float:
        """Returns an ini file entry as a floating point number."""
        key = f"{section}/{entry}"
        if key in self._entries:
            try:
                return float(self._entries[key])
            except ValueError:
                return default_value
        return default_value

    def get_int(self, section: str, entry: str, default_value: int) -> int:
        """Returns an ini file entry as an integer number."""
        key = f"{section}/{entry}"
        if key in self._entries:
            try:
                return int(self._entries[key])
            except ValueError:
                return default_value
        return default_value

    def get_string(self, section: str, entry: str, default_value: str) -> str:
        """Returns an ini file entry as a string."""
        key = f"{section}/{entry}"
        return self._entries.get(key, default_value)


class WriteIniFile:
    def __init__(self):
        self._entries = {} # type: dict[str, dict[str, str]]

    def add_section(self, section: str):
        if section not in self._entries:
            self._entries[section] = {}

    def put_float(self, section: str, entry: str, value: float) -> bool:
        if section not in self._entries:
            return False
        # Matches C++ snprintf(buffer, 16, "%f", value)
        self._entries[section][entry] = f"{value:f}"
        return True

    def put_int(self, section: str, entry: str, value: int) -> bool:
        if section not in self._entries:
            return False
        self._entries[section][entry] = str(value)
        return True

    def put_string(self, section: str, entry: str, value: str) -> bool:
        if section not in self._entries:
            return False
        self._entries[section][entry] = value
        return True

    def write(self, config_file: str) -> None:
        try:
            with open(config_file, 'w', encoding='utf-8') as f:
                for section_name, section_data in self._entries.items():
                    f.write(f"[{section_name}]\n")
                    for key, value in section_data.items():
                        f.write(f"{key}={value}\n")
                    f.write("\n")
        except IOError:
            pass

File: groundfire-0.25/src/inifile.cc
////////////////////////////////////////////////////////////////////////////////
//
//               Groundfire
//
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2004, Tom Russell (tom@groundfire.net)
//
// This file is part of the Groundfire project, distributed under the MIT 
// license. See the file 'COPYING', included with this distribution, for a copy
// of the full MIT licence.
//
////////////////////////////////////////////////////////////////////////////////
//
//   File name : inifile.cc
//
//          By : Tom Russell
//
//        Date : 25-Feb-04
//
// Description : Generic module to handle reading and writing ini files. 
//
//               NOTE : The writing of ini files is currently untested!
//
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// Includes
////////////////////////////////////////////////////////////////////////////////
#include <fstream>
#include <stdio.h>
#ifdef DEBUG
#include <windows.h>
#endif
#include "inifile.hh"

////////////////////////////////////////////////////////////////////////////////
// Public Member Functions
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
//
// Function    : cReadIniFile
//
// Description : Constructor
//
////////////////////////////////////////////////////////////////////////////////
cReadIniFile::cReadIniFile
(
    string const & configFile
) 
{
    ifstream file (configFile.c_str ());

    if (!file)
    {
        return;
    }
    
    string line;
    string name;
    string value;
    string inSection;
    int    posEqual;

    while (getline (file, line))
    {
        if (!line.length ()) continue;
        
        // Ignore comment lines
        if (line[0] == '#') continue;
        if (line[0] == ';') continue;
        
        // A new section, record it.
        if (line[0] == '[')
        {
            inSection = line.substr (1, line.find (']') - 1);
            continue;
        }
        
        // get a value
        posEqual = line.find ('=');

        name  = line.substr (0, posEqual);
        value = line.substr (posEqual + 1);

        // Strip whitespace from the beginning and end of the strings
        name.erase  (0, name.find_first_not_of ("\t "));
        name.erase  (   name.find_last_not_of ("\t ") + 1);
        value.erase (0, value.find_first_not_of ("\t "));
        value.erase (   value.find_last_not_of ("\t ") + 1);

#ifdef DEBUG
        OutputDebugString (inSection.c_str ());
        OutputDebugString (name.c_str ());
        OutputDebugString (value.c_str ());
        OutputDebugString ("\n");
#endif
        
        // Insert this entry into the map.
        _entries[inSection + '/' + name] = value;
    }
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : getFloat
//
// Description : Returns an ini file entry as a floating point number
//
////////////////////////////////////////////////////////////////////////////////
float
cReadIniFile::getFloat
(
    string const & section,
    string const & entry,
    float          defaultValue
) 
const
{
    map<string,string>::const_iterator ci 
        = _entries.find (section + '/' + entry);
    
    if (ci == _entries.end ()) 
    {
        return (defaultValue);
    }

    return ((float)atof (ci->second.c_str ()));
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : getInt
//
// Description : Returns an ini file entry as a integer number.
//
////////////////////////////////////////////////////////////////////////////////
int
cReadIniFile::getInt
(
    string const & section,
    string const & entry,
    int            defaultValue
) 
const
{
    map<string,string>::const_iterator ci 
        = _entries.find (section + '/' + entry);
    
    if (ci == _entries.end ()) 
    {
        return (defaultValue);
    }

    return (atoi (ci->second.c_str ()));
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : getString
//
// Description : Returns an ini file entry as a string.
//
////////////////////////////////////////////////////////////////////////////////
string
cReadIniFile::getString
(
    string const & section,
    string const & entry,
    string         defaultValue
) 
const
{
    map<string,string>::const_iterator ci 
        = _entries.find (section + '/' + entry);
    
    if (ci == _entries.end ()) 
    {
        return (defaultValue);
    }
    
    return (ci->second);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : addSection
//
// Description : Adds a new section to an ini file.
//
////////////////////////////////////////////////////////////////////////////////
void
cWriteIniFile::addSection
(
    string const & section
)
{
    map<string, string> newEntry;
    _entries[section] = newEntry;
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : putFloat
//
// Description : Puts a floating point value entry into the ini file
//
////////////////////////////////////////////////////////////////////////////////
bool
cWriteIniFile::putFloat
(
    string const & section,
    string const & entry,
    float          value
)
{
    map<string,map<string, string> >::iterator ci = _entries.find (section);

    if (ci == _entries.end ())
    {
        return (false);
    }

    char buffer[16];
    snprintf (buffer, 16, "%f", value);

    ci->second[entry] = buffer;

    return (true);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : putInt
//
// Description : Puts an integer value entry into the ini file
//
////////////////////////////////////////////////////////////////////////////////
bool
cWriteIniFile::putInt
(
    string const & section,
    string const & entry,
    int            value
)
{
    map<string,map<string, string> >::iterator ci = _entries.find (section);

    if (ci == _entries.end ())
    {
        return (false);
    }

    ci->second[entry] = value;

    return (true);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : putString
//
// Description : puts a string value into the ini file
//
////////////////////////////////////////////////////////////////////////////////
bool
cWriteIniFile::putString
(
    string const & section,
    string const & entry,
    string const & value
)
{
    map<string,map<string, string> >::iterator ci = _entries.find (section);

    if (ci == _entries.end ())
    {
        return (false);
    }

    ci->second[entry] = value;

    return (true);    
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : write
//
// Description : Writes out the ini file to disk.
//
////////////////////////////////////////////////////////////////////////////////
void 
cWriteIniFile::write 
(
    string const & configFile // The file name of the ini file.
)
{
    ofstream file (configFile.c_str ());

    map<string, map<string, string> >::iterator i;

    for (i = _entries.begin (); i != _entries.end (); i++)
    {
        file << "[" << i->first << "]" << endl;

        map<string, string>::iterator j;

        for (j = i->second.begin (); j != i->second.end (); j++)
        {
            file << j->first << "=" << j->second << endl; 
        }

        file << endl;
    }
}
